### DISCUSSION TITLE
A way to prevent `nvim-tree` buffers in shada

### DISCUSSION THREAD

**Discussion by litoj** [nvim-tree/nvim-tree.lua#2365] [General]
I would like to unclutter my history from all the `/path/to/cwd/NvimTree_1` files, that never existed, but neovim still saves those paths in my shada, therefore lowering my history of actual files.

Unfortunately, from what I know, it is not possible to use `vim.o.shada = 'r:.*/NvimTree_1'.`

Is there a way to change it to something more like terminal notation, that can be caught by shada pattern? (`term:`...)

**Comment by alex-courtis** [comment]
That is indeed a problem I was not aware of.

```
:; cat m.shada | grep -E "^[^ ]" | sed -E 's/with timestamp.*//g' | sort | uniq -c
    912 Change
     12 Global mark
      1 Header
   9317 History entry
    100 Jump
    155 Local mark
     37 Register
      1 Replacement string
      2 Search pattern
```

It's no so bad:
```
:; cat m.shada | grep NvimTree_1 | wc -l
271
:; cat m.shada | grep fugitive | wc -l
31
:; cat m.shada | grep --no-group-separator -B 2 NvimTree_1 | grep -E "^[^ ]" | sed -E 's/with timestamp.*//g' | sort | uniq -c
213 Change
25 Jump
32 Local mark
```

**Comment by alex-courtis** [comment]
I'm not well versed in shada and can't see any doc around the pattern.

It seems that notation like `fugitive://` isn't automagically applied.

I'm open to any and all ideas...

**Comment by litoj** [comment]
it is not automatic, but any prefixes can be filtered by `vim.opt.shada:append("r<prefix>")`, so if the name of NvimTree was also in some prefix notation, it would allow for users to fix the issue.

**Reply by alex-courtis** [reply]
I see. It works for `vim.opt.shada:append("rfugitive://")` but not the example from the help `vim.opt.shada:append("r/tmp")`

An `nvim-tree://` or similar prefix would work.

We use the name `NvimTree_<tabpage>` however that could be changed to something like `nvim-tree://NvimTree_<tabpage>`. We cannot break users who match `NvimTree_<tabpage>` in their own plugins and scripts.

I'm happy to make such a change but we need to thoroughly test it first. Are you willing to change this?

**Reply by litoj** [reply]
I am unsure what change are you referring to, but sure. I would like to help in any way possible.

**Reply by alex-courtis** [reply]
Fantastic. What change are you unsure of?

**Reply by litoj** [reply]
Well, what you meant by 'willing to change this', but I assume a PR with the adapted code was it.

**Reply by alex-courtis** [reply]
Yes please: [CONTRIBUTING.md](https://github.com/nvim-tree/nvim-tree.lua/blob/master/CONTRIBUTING.md)

[wiki: Development](https://github.com/nvim-tree/nvim-tree.lua/wiki/Development)

**Comment by hinell** [comment]
Few options that come into my mind:
* forcibly close all `NvimTree` windows just before exiting nvim
* provide full paths to the `shada+=r/path/to/NvimTree_N` just before exiting nvim
* open a tracking issue in [upstream nvim repo](https://github.com/neovim/neovim/issues) to requestion feature to exclude non-writable buffers' with fake files from shada

**Reply by litoj** [reply]
There are multiple issues opened and set as `backlog` in the repo, so I think the way shada works won't change any time soon.

I haven't thought about the way of adding extra paths to shada before write, that is very clever and solves the issue perfectly. I ended up with code like this:
```lua
vim.api.create_autocmd('VimLeavePre', {callback = function()
	local filter = { vim.o.shada }
	for _, v in ipairs(vim.v.oldfiles) do
		if v:find('.git/', 7, true) or v:match 'NvimTree_%d+$' or not vim.loop.fs_stat(v) then
			filter[#filter + 1] = v
		end
	end
	vim.o.shada = table.concat(filter, ',r')
end})

-- in config: 'wipeout' doesn't add windows to the 'jumps' list
vim.keymap.set('n', 'q', vim.cmd.bwipeout, {buffer = true})
```
Thank you so much for the idea!

**Reply by hinell** [reply]
I've actually came up with  a bit more efficient approach. It excludes every `NvimTree_N` path from `:oldfiles` list:

```lua
M = {}
M.shada_exclude = {}
vim.api.nvim_create_augroup("NvimTreeBufferLocal", { clear = true })
vim.api.nvim_create_autocmd({
	"FileType"
},{
	pattern  = "NvimTree_*",
	group    = "NvimTreeBufferLocal",
	callback = function()

		local bufName    = vim.api.nvim_buf_get_name(auEvent.buf)
		if not M.shada_exclude[tabCurrent] then
			M.shada_exclude[tabCurrent] = bufName
			vim.opt.shada:append("r" .. bufName)
		end

	end
})

return M
```

**Reply by hinell** [reply]
Calling `vim.loop.fs_stat(v)` is huge overkill as there are hundreds of files.

**Reply by litoj** [reply]
I absolutely agree testing file existence is overkill, for nvimtree the regex suffices. I think your solution would have to be altered slightly to exclude every nvimtree window per cwd. The cwd is the main hurdle. The tabs can be safely ignored altogether because shada matches by prefix, not full path.
